### DISCUSSION TITLE
Ganged Axis / Squaring and a few other questions on STM32F4

### DISCUSSION THREAD

**Discussion by rvalotta** [terjeio/grblHAL#307] [Q&A]
I have spent the last 2 days working out how to get Ganged Axis, Squaring implemented in the STM32F4 driver.    I seem to have managed to make it work. The driver appears to allow you to do a mapped io or a bit band in the step and direction functions.  Using the bitband section I've added the appropriate #if.... and have it working nicely.   I think I finally understand how the mapped io works but it looks like its going to want me to modify grbl/stepdir_map c_step_outmap to include the xyz2 bits and similar for direction.    Looking for some feedback on if its worth doing. I'm guessing the time it takes to set all the bits at once vs each step/dir individually might not be that significant.  

My questions.... I am going to guess that the squaring limit pin does not get reported in the console in any way. Is this correct?

**Comment by rvalotta** [comment]
Just looking at the arrays in stepdir_map.h and all i see is a nightmare of #if x2.... if y2... if z2... for every line that has a x y or z.

**Reply by terjeio** [reply]
> I am going to guess that the squaring limit pin does not get reported in the console in any way. Is this correct?

Yes and no - in the real-time report thery are combined when outputting the per pin state. Use `$LIM` to output a report with individual switches listed.

>  Using the bitband section I've added the appropriate #if.... and have it working nicely.

Good.

>  I think I finally understand how the mapped io works but it looks like its going to want me to modify grbl/stepdir_map c_step_outmap to include the xyz2 bits and similar for direction. Looking for some feedback on if its worth doing.

You may use mapped io for the main outputs, always use bitbanding for *2 outputs. Usually there is only one linked motor so it does not make sense to use mapping for that.

> I'm guessing the time it takes to set all the bits at once vs each step/dir individually might not be that significant.

True for most applications, if laser engraving bitmaps at very high speed it may become significant.
